openstreetmap-carto.lua

```diff
@@ -303,134 +577,169 @@ function filter_tags_generic(tags)
-
-    -- Start by assuming that this not an old-style MP
-    for i = 1, membercount do
-        members_superseded[i] = 0
-    end
+function add_roads(object)
+    local cols = split_tags(object.tags, columns_map.roads)
+    cols['z_order'] = z_order(object.tags)
+    cols.way = { create = 'line', split_at = 100000 }
+    tables.roads:add_row(cols)
+end
 
-    local type = keyvalues["type"]
+function add_polygon(object)
+    local cols = split_tags(object.tags, columns_map.polygon)
+    cols['z_order'] = z_order(object.tags)
+    cols.way = { create = 'area', split_at = nil }
+    tables.polygon:add_row(cols)
+end
 
-    -- Remove type key
-    keyvalues["type"] = nil
+function add_transport_polygon(object)
+    local cols = split_tags(object.tags, columns_map.transport_polygon)
+    cols['z_order'] = z_order(object.tags)
+    cols.way = { create = 'area', split_at = nil }
+    tables.transport_polygon:add_row(cols)
+end
 
-    -- Filter out relations with just a type tag or no tags
-    if next(keyvalues) == nil then
-        return 1, keyvalues, members_superseded, 0, 0, 0
+function add_route(object)
+    for i, member in ipairs(object.members) do
+        if member.type == 'w' then
+            local cols = split_tags(object.tags, columns_map.route)
+            cols.member_id = member.ref
+            cols.member_position = i
+            tables.route:add_row(cols)
+        end
     end
+end
 
-    if type == "boundary" or (type == "multipolygon" and keyvalues["boundary"]) then
-        keyvalues.z_order = z_order(keyvalues)
-        return 0, keyvalues, members_superseded, 1, 0, roads(keyvalues)
-    -- For multipolygons...
-    elseif (type == "multipolygon") then
-        -- Multipolygons by definition are polygons, so we know roads = linestring = 0, polygon = 1
-        keyvalues.z_order = z_order(keyvalues)
-        return 0, keyvalues, members_superseded, 0, 1, 0
-    elseif type == "route" then
-        keyvalues.z_order = z_order(keyvalues)
-        return 0, keyvalues, members_superseded, 1, 0, roads(keyvalues)
+function osm2pgsql.process_node(object)
+    if clean_tags(object.tags) then
+        return
     end
 
-    -- Unknown type of relation or no type tag
-    return 1, keyvalues, members_superseded, 0, 0, 0
+    add_point(object)
 end
 
---- Check if an object with given tags should be treated as polygon
--- @param tags OSM tags
--- @return 1 if area, 0 if linear
-function isarea (tags)
-    -- Treat objects tagged as area=yes polygon, other area as no
-    if tags["area"] then
-        return tags["area"] == "yes" and 1 or 0
+function osm2pgsql.process_way(object)
+    if osm2pgsql.stage == 2 then
+        -- Stage two processing is called on ways that are part of admin boundary relations
+        local props = phase2_admin_ways[object.id]
+        if props ~= nil then
+            tables.admin:add_row({admin_level = props.level, multiple_relations = (props.parents > 1), geom = { create = 'line' }})
```

Comment:
The

`geom = { create = 'line' }`

part is not an ordinary column definition, it is a "geometry transformation", as part of the "processing callbacks" you referred to. It is probably an osm2pgsql thing needing a little bit better documentation, but I think the "geom" part is just a internal osm2pgsql keyword here, used to refer to the single geometry column that AFAIK is currently supported. If I remember it well, Jochen was also exploring the possibility of having multiple geometry columns in a single table, so this particular thing might actually change in the future.

This code definitely works by the way in its current form. I have used it extensively to import data with success while exploring and testing Paul's work.

Reply:
I think you're right. See https://osm2pgsql.org/doc/manual.html#geometry-transformations

This works right now because for way tables, the default is `{ create = 'line' }`